sw/device/lib/base/memory_hardened.c

```diff
@@ -0,0 +1,703 @@
+static uintptr_t select_addr(uintptr_t mask, uintptr_t addr_ones,
+                             void *addr_zeroes) {
+  return (mask & addr_ones) | (~mask & (uintptr_t)addr_zeroes);
+}
+
+// Memory copy with shuffled memory access pattern
+hardened_bool_t shuffled_memcpy(void *restrict dest, const void *restrict src,
+                                size_t len, uint32_t seed) {
+  uintptr_t dest_addr = (uintptr_t)dest;
+  uintptr_t src_addr = (uintptr_t)src;
+  uintptr_t tail = dest_addr + len;
+  // Use volatile to reload tail_copy and not reduce math
+  volatile uintptr_t tail_copy = tail;
+
+  // Simple protection against reusing seed.
+  seed ^= dest_addr + (src_addr << 9);
+
+  // Decoy memory for copy.
+  uintptr_t decoy1 = seed;
+  uintptr_t decoy2 = dest_addr;
+
+  // Calculate how many heading bytes should be processed.
+  size_t bytes_to_process = len;
+  if (((dest_addr & WORD_MASK) == (src_addr & WORD_MASK)) &&
+      (len >= sizeof(uintptr_t))) {
+    // Set 'head' to the first word boundary if src & dest can be aligned.
+    bytes_to_process = (-dest_addr) & WORD_MASK;
+  }
+
+  // Remaining part can be processed word aligned.
+  size_t words_to_process = (len - bytes_to_process) / sizeof(uintptr_t);
+
+  // Shuffle at byte level.
+  while (bytes_to_process) {
+    size_t n_min = min(bytes_to_process, PRIME_MAX);
+    size_t n_prime = larger_prime(n_min);
+
+    // Select random starting position and stride.
+    size_t pos = div_by_mul(n_prime, seed);
+    seed = next_fast_random(seed);
+    size_t stride = div_by_mul(n_prime - 4, seed) + 3; /* n_prime >= 11 */
+
+    size_t processed = 0;
+    while (processed < n_min) {
+      uintptr_t mask = -(pos < n_min); /* copy if not a decoy */
+      uintptr_t saddr = select_addr(mask, src_addr + pos, &decoy1);
+      uintptr_t daddr = select_addr(mask, dest_addr + pos, &decoy2);
+      processed -= mask; /* increment count if (start < n_min) */
+      *((volatile uint8_t *)daddr) = *((uint8_t *)saddr);
+      pos += stride;
+      /* pos = pos mod n_prime */
+      pos -= n_prime & (-(pos >= n_prime));
+    }
+    dest_addr += n_min;
+    src_addr += n_min;
+    bytes_to_process -= n_min;
+  }
+
+  while (words_to_process) {
+    // Copy in chunks of PRIME_MAX words.
+    size_t n_min = min(words_to_process, PRIME_MAX);
+    size_t n_prime = larger_prime(n_min);
+
+    // Select random starting position and stride.
+    size_t pos = div_by_mul(n_prime, seed);
+    seed = next_fast_random(seed);
+    size_t stride = div_by_mul(n_prime - 4, seed) + 3;  // n_prime >= 11
+
+    size_t processed = 0;
+    while (processed < n_min) {
+      uintptr_t mask = -(pos < n_min);  // Copy if not a decoy.
+      uintptr_t pos_word = pos * sizeof(uintptr_t);
+      uintptr_t saddr = select_addr(mask, src_addr + pos_word, &decoy1);
+      uintptr_t daddr = select_addr(mask, dest_addr + pos_word, &decoy2);
+      processed -= mask;  // increment count if (start < n_min)
```

Comment:
Can you please carefully go through the CL and ensure that all comments are of `//`, begin with a capital letter, and end with a period? I think you missed a few.